### PR TITLE
Handle edge case with stats calculation

### PR DESCRIPTION
**Pull request by dgonzdev**
pr for this ticket https://github.com/nycmeshnet/meshdb/issues/950

the code creates an array of size 100, where the index represents a time period between the start_time and end_time parameters, and the value represents the total number of active installs or all installs during that time period.

we retrieve the installs to populate the counts in the array with these queries
- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L65
- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L71

they filter for installs between the start_time and end_time with some other conditional params

it's currently breaking when an edge case causes us to index into the array out of range.

i believe there are three situations where that can occur

- we use the request date on an install instead of the install date and the request date is before start_time (i don't think this would actually occur b/c how can install date every be null if we are filtering based on it in the db query?)
-- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L79

- the request date is the same as the end_time parameter which would cause is to index into the array at index 100
-- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L84

- the install date is the same as the end_time parameter which would cause is to index into the array at index 100
-- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L84

my fix is to index into the array at the last position 99 when the request date or install date are the same as the end_time parameter

i also separated the code paths for generating buckets for active installs vs all installs to make it clearer how the queries for each worked at the expense of keeping the code "dry".

this is super edge case-y. happy to write a test if needed.

also, i could be wrong and the error could be caused by something else.

### CODE REVIEW

File: src/meshweb/views/website_stats.py [hunk 1]
```diff
@@ -53,36 +53,79 @@ def cors_allow_website_stats_to_all(sender: None, request: HttpRequest, **kwargs
...
+    object_queryset = base_object_queryset.filter(
+        install_date__gte=start_datetime.date(),
+        install_date__lte=end_datetime.date(),
+    )
+    buckets[0] = base_object_queryset.filter(install_date__lt=start_datetime).count()
 
     for install in object_queryset:
-        if data_source == "active_installs":
-            counting_date = install.install_date or install.request_date.date()
+        counting_date = install.install_date
```
[the remark this answers is not in the document]
Comment (dgonzdev) [reply src/meshweb/views/website_stats.py:80]: i added a conditional to continue at line 82 and not count the install if it has no install_date

if the install has no install_date, is it active?

[the remark this answers is not in the document]
Comment (dgonzdev) [reply src/meshweb/views/website_stats.py:80]: yes, it does.

existing all case implementation:
- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L71

existing active only implementation:
- https://github.com/nycmeshnet/meshdb/blob/main/src/meshweb/views/website_stats.py#L65

[the remark this answers is not in the document]
Comment (Andrew-Dickinson) [reply src/meshweb/views/website_stats.py:80]: Looks like yes. https://github.com/nycmeshnet/meshdb/blob/d7543096915623c319878f54a8019674fae5eafd/src/meshweb/views/website_stats.py#L190

All good, let’s add a regression test here and we can ship the logic as is

**Comment by dgonzdev** [comment]
this probably needs more refactoring but going to stop here.

i think this was breaking when we had an install at the same time as the request was happening.

the bucket is an array with 0-99 indexes where each index is a count of the number of installs in the time period that index represents.

we do different db queries to retrieve active or requested installs. and we were doing a less than or equal for the upper bound time range against a date (format year-month-day).

if there was an install with an active date or requested date that was equal to the upper bound, the math that determines what index in the bucket it should count to would return an index of 100 which is out of range. this is more likely to happen with active date's b/c those are stored as dates (requested dates are stored as timestamps and include seconds).

i made a couple changes separating parts of the code paths for the active vs requested install queries, making the upper bound less than instead of less than or equal, and adding a check to continue if the index is 100.

the bigger refactor would probably be redoing the queries and bucket math so an out of range is never even possible (plus maybe storing active installs as timestamps with seconds instead of just dates?)

**Comment by dgonzdev** [comment]
fixed the linting error and test error when the bucket index was zero
